### PR TITLE
gh-116810: fix memory leak in ssl module

### PR DESCRIPTION
**Pull request by jeffvanvoorst**
PySSL_get_session leaks a session object each time it is called.  For programs like cherrypy (cheroot), a leak occurs with each handled web request.

* Issue: gh-116810

### CODE REVIEW

File: Misc/NEWS.d/next/Library/2024-08-23-15-49-10.gh-issue-116810.QLBUU8.rst
Comment (ZeroIntensity) [inline Misc/NEWS.d/next/Library/2024-08-23-15-49-10.gh-issue-116810.QLBUU8.rst]:
```suggestion
Resolved SSL session memory leak in :mod:`ssl`.
```

File: Modules/_ssl.c
Comment (webknjaz) [inline Modules/_ssl.c]: Judging by no test failures, this probably does not have any coverage. Any chance of including the tests in this PR?

Comment (jeffvanvoorst) [reply Modules/_ssl.c]: I can check the tests this week.  I doubt I will be able to finish that work today.

Comment (jeffvanvoorst) [reply Modules/_ssl.c]: Since this change still results in returning a valid copy of the ssl session, I doubt tests would fail.

Comment (ZeroIntensity) [reply Modules/_ssl.c]: Yeah, agreeing there. I think it would be difficult to test this. With that being said, are there any existing tests for checking that the `session` attribute is valid? If not, then this PR should add them (otherwise, I think this looks good).

Comment (jeffvanvoorst) [reply Modules/_ssl.c]: I had some life issues come up.  I will try to get back to this in the next few weeks.

Comment (ZeroIntensity) [reply Modules/_ssl.c]: No worries, thanks for contributing!

**Comment by vstinner** [comment]
I confirm that **the change works as expected**: it fixes a significant memory leak.

I modified the reproducer to dump the RSS memory usage on Linux:

```py
import os
import socket
import ssl
import gc

host = '142.251.163.99' # some server we can connect with https as example
port = 443

session = None
context = ssl._create_unverified_context(protocol=ssl.PROTOCOL_TLSv1_2)
with socket.create_connection((host, port)) as sock:
    ssock = context.wrap_socket(sock, server_hostname=host, session=session)
    with ssock:
        for i in range(300000):
            session = ssock.session
            if (i % 1000) == 0:
                gc.collect()
                os.system(f"grep ^VmRSS /proc/{os.getpid()}/status")
```

I can easily reproduce the bug:

```
VmRSS:	   19840 kB
VmRSS:	   26496 kB
VmRSS:	   33280 kB
VmRSS:	   40064 kB
VmRSS:	   46848 kB
VmRSS:	   53632 kB
VmRSS:	   60416 kB
VmRSS:	   67200 kB
VmRSS:	   74112 kB
VmRSS:	   80896 kB
VmRSS:	   87680 kB
VmRSS:	   94464 kB
VmRSS:	  101248 kB
(...)
```

With the PR, there is no leak anymore:

```
VmRSS:	   20436 kB
VmRSS:	   20436 kB
VmRSS:	   20436 kB
VmRSS:	   20436 kB
VmRSS:	   20436 kB
VmRSS:	   20436 kB
VmRSS:	   20436 kB
VmRSS:	   20436 kB
VmRSS:	   20436 kB
VmRSS:	   20436 kB
VmRSS:	   20436 kB
VmRSS:	   20436 kB
VmRSS:	   20436 kB
VmRSS:	   20436 kB
VmRSS:	   20436 kB
(...)
```

**Review by vstinner — APPROVED** [review]
LGTM

**Comment by vstinner** [comment]
@gpshead @pitrou: Do you want to review this change? (I don't recall if @pitrou worked on `ssl` or not, ignore me if I'm wrong.)

**Comment by pitrou** [comment]
Yes, I worked on `ssl`, but that was long ago.

File: Misc/NEWS.d/next/Library/2024-08-23-15-49-10.gh-issue-116810.QLBUU8.rst
Comment (pitrou) [inline Misc/NEWS.d/next/Library/2024-08-23-15-49-10.gh-issue-116810.QLBUU8.rst]:
```suggestion
Resolve SSL session memory leak in :mod:`ssl`.
```

File: Modules/_ssl.c
Comment (pitrou) [reply Modules/_ssl.c]: I don't think it's possible to test anything, other than manually checking the leak is fixed as @vstinner did.

The issue was apparently that we were first duplicating the session using `_ssl_session_dup`, but then we immediately forgot the duplicated session by calling `SSL_get1_session`.

Comment (pitrou) [reply Modules/_ssl.c]: That said, a potential alternate fix would be to remove the `SSL_get0_session` + `_ssl_session_dup` dance and keep the `SSL_get1_session` call. @jeffvanvoorst Can you try to see if it works?

Comment (gpshead) [reply Modules/_ssl.c]: Based on my understanding of https://docs.openssl.org/master/man3/SSL_get_session/ I believe this will work, but it does raise the question of _why_ we have the `SSL_get0_session + _ssl_session_dup` combo in the first place; with modern OpenSSL 3+ why would this `SSL_get1_session` not just do what we wanted? @sethmlarson

Comment (pitrou) [reply Modules/_ssl.c]: I'll also note that the issue linked to in the code above was finally [closed](https://github.com/openssl/openssl/issues/1550#issuecomment-287164456) claiming it's a Python issue.

Comment (gpshead) [reply Modules/_ssl.c]: I think the issue @.davidben was saying Python has in there was describing more the way Python is interacting with the available APIs (which at the time had to span a wide range of OpenSSL versions) and not using some other features instead.

Comment (vstinner) [reply Modules/_ssl.c]: https://docs.python.org/dev/using/configure.html#build-requirements says:
> OpenSSL 1.1.1 is the minimum version and OpenSSL 3.0.9 is the recommended minimum version for the ssl and hashlib extension modules.

Does OpenSSL 1.1.1 need the _ssl_session_dup() workaround?

Comment (gpshead) [reply Modules/_ssl.c]: For 3.13 I don't think we care about 1.1.1 anymore given it is past EOL. https://discuss.python.org/t/pathway-to-requiring-openssl-3-or-newer-in-cpython/46055/10

Comment (gpshead) [reply Modules/_ssl.c]: The OpenSSL 1.1.1 fork's I do think we should care somewhat about are BoringSSL and it's AWS-LC derivative. @WillChilds-Klein and @davidben can represent what those might need for these specific APIs.

Comment (gpshead) [reply Modules/_ssl.c]: I don't really understand how this code was nor always a memory leak in the past. It implies that the `SSL_get0_session + _ssl_session_dup()` logic above combined with a subsequent `SSL_get1_session` call were somehow not returning two distinct things or doubly increasing a reference count which by my read I think they should have been? (and appear to be now in 3.x)

Comment (davidben) [reply Modules/_ssl.c]: There's a cacophony of mistakes here on the CPython side. Let me try to correct things:

> I don't really understand how this code was nor always a memory leak in the past.

The memory leak seems to have been introduced by https://github.com/python/cpython/commit/39258d3595300bc7b952854c915f63ae2d4b9c3e. `_ssl_session_dup` returns an `SSL_SESSION` that needs to be freed by someone and then you all just immediately clobber the variable with something else. But `test_ssl.py` clearly accesses `session`. Does CPython have no way to catch memory leaks in its tests? Might I suggest you all look into LeakSanitizer?

Anyway, the problem with that commit is that you all turned `#if ... #else ... #endif` into including both branches instead of just one.

That commit not only introduced a memory leak, but it also mostly broke your workaround. I haven't tried it, but I suspect you would see this if you _uncleanly_ destroyed the other connection _before_ calling `set_session` instead of after. Now, I see _uncleanly_ because it seems CPython never followed up on https://github.com/openssl/openssl/issues/1550#issuecomment-287179654 because you all have been misunderstanding the nature of the workaround. It was not working around an OpenSSL bug, but a CPython bug / broken expectation.

> Based on my understanding of https://docs.openssl.org/master/man3/SSL_get_session/ I believe this will work, but it does raise the question of why we have the SSL_get0_session + _ssl_session_dup combo in the first place; with modern OpenSSL 3+ why would this SSL_get1_session not just do what we wanted?

That is not what is going on here. There were, as far as I know, no changes in OpenSSL 3+ relating to this. 

`SSL_get0_session + _ssl_session_dup` and `SSL_get1_session` do not and have never same thing in any version of OpenSSL. `SSL_get0_session` and `SSL_get1_session` are the exact same accessor. Just one of them increments the refcount for you and the other doesn't. `_ssl_session_dup` is a (misguided) CPython function to make a deep copy of the `SSL_SESSION` object.

Although the code was, prior to the regression, gated on a `OPENSSL_VERSION_1_1` ifdef, that does _not_ imply was specific to OpenSSL 1.1.x. `OPENSSL_VERSION_1_1` is a CPython-invented macro (that squats OpenSSL's namespace) which means > 1.1. That is, OpenSSL 3 was also `OPENSSL_VERSION_1_1`:

```
#if (OPENSSL_VERSION_NUMBER >= 0x10100000L) && !defined(LIBRESSL_VERSION_NUMBER)
#  define OPENSSL_VERSION_1_1 1
#  define PY_OPENSSL_1_1_API 1
#endif
```

While I have not tested it (have you all?), based on https://github.com/openssl/openssl/issues/1550#issuecomment-287179654, I highly suspect that CPython "needs" the workaround in OpenSSL 3.x too. I say "needs" because...

> I think the issue @.davidben was saying Python has in there was describing more the way Python is interacting with the available APIs (which at the time had to span a wide range of OpenSSL versions) and not using some other features instead.

No, that's not what I was saying in https://github.com/openssl/openssl/issues/1550#issuecomment-287179654. See the immediately preceding comment:

> FYI, I poked at this briefly and don't see any evidence that SSL_shutdown is being called by Python. It seems the pseudocode in the bug report is wrong and this is a Python bug.

The bug report was very confused because the minimal reproducer that CPython provided did not actually reflect what CPython does. With all the meandering with the bad reproducer, I suspect it was very confusing. But the CPython folks on the bug never followed up, so my past self assumed you all understood, were planning on fixing it, and I didn't have time to check on you all. 😞

What is _actually_ going on here is that OpenSSL will invalidate sessions on connection destruction if you haven't called `SSL_shutdown` first. To be honest, I don't think this is good behavior, but it has been OpenSSL's intentional behavior since day 1. I believe this is descended from this guidance:

>    close_notify
>       This message notifies the recipient that the sender will not send
>       any more messages on this connection. The session becomes
>       unresumable if any connection is terminated without proper
>       close_notify messages with level equal to warning.

https://www.rfc-editor.org/rfc/rfc2246.html#section-7.2.1

Now, I personally think this behavior is unhelpful. If a connection just got closed, that doesn't make the old key material broken or anything. Also if we were getting any security properties out of this, it wouldn't work anyway. There are loads of cases where you cannot reliably knock out the other session.

* Stateless resumption ("session tickets" in TLS 1.2) cannot possibly invalidate the session on the server
* Maybe your application serializes the session, e.g., on disk, and then deserializes it later

Interestingly, OpenSSL doesn't actually look for a full bidirectional shutdown to preserve your session. It only looks for whether you attempted to *send* `close_notify`. Kind of pointless, to be honest, but IMO this whole thing is pointless.

TLS 1.1, incidentally, already softened this recommendation:

>    close_notify
>      This message notifies the recipient that the sender will not send
>      any more messages on this connection.  Note that as of TLS 1.1,
>      failure to properly close a connection no longer requires that a
>      session not be resumed.  This is a change from TLS 1.0 to conform
>      with widespread implementation practice.

https://www.rfc-editor.org/rfc/rfc4346.html#section-7.2.1

But OpenSSL retains this requirement, for better or worse. This requirement was not new in 1.1.x, nor was it removed in 3.x. The only reason you all needed to work around it later on is that, due to other quirks of how CPython was using the API (arguably incorrectly), as detailed in my comment on the OpenSSL issue, CPython was sneaking through a gap in this OpenSSL policy. OpenSSL fixed that gap, and now CPython was affected.

Now, contrary to the claim in https://github.com/openssl/openssl/issues/1550#issue-175600961, CPython did not call `SSL_shutdown`. You all have a separate method to exchange close_notify and weren't calling it in that test.

So, what to do given this? I think you have two options:

1. Agree with OpenSSL's policy and pass it up to the callers. Without calling `SSLSocket.shutdown()`, sessions get invalidated. Callers have to call the shutdown method if they care. Fix your tests to reflect this.
2. Disagree with OpenSSL's policy and effectively turn it off. To do that, call `SSL_shutdown` when destroying an `SSL`. Now, that will cause OpenSSL to do I/O, so you might want to `SSL_set_quiet_shutdown`. Alternatively, there is an `SSL_set_shutdown` API to just fake a shutdown state.

In either case, you can remove the workaround because this workaround has been self-inflicted from day one.

If you pick option 2, please write some clear comments in the code so the next time this happens, I don't have to write another essay for you all. :-P

> The OpenSSL 1.1.1 fork's I do think we should care somewhat about are BoringSSL and it's AWS-LC derivative. @WillChilds-Klein and @davidben can represent what those might need for these specific APIs.

So, per the above, this question is based on several incorrect premises:
* BoringSSL is not an OpenSSL 1.1.1 fork. We actually diverged from OpenSSL 1.0.2. We just report a version of 1.1.1 because that's what's most convenient for compatibility.
* This issue has nothing to do with OpenSSL 1.1.1. I expect you all still "need" this workaround in OpenSSL 3.x too.
* This issue has nothing to do with these specific APIs

Since this "workaround" is really a workaround for CPython's own mishaps, if you all do the thing I suggest above, it should work just fine in BoringSSL.

But also this is all moot in BoringSSL because BoringSSL does not invalidate sessions on `SSL_free` in the first place. We removed it because it's just kinda silly.

Comment (davidben) [reply Modules/_ssl.c]: This is where OpenSSL documents their behavior, by the way:

> A session will be automatically removed from the session cache and marked as non-resumable if the connection is not closed down cleanly, e.g. if a fatal error occurs on the connection or [SSL_shutdown(3)](https://docs.openssl.org/master/man3/SSL_shutdown/) is not called prior to [SSL_free(3)](https://docs.openssl.org/master/man3/SSL_free/).

https://docs.openssl.org/master/man3/SSL_get_session/#notes:~:text=A%20session%20will%20be%20automatically%20removed%20from%20the%20session%20cache%20and%20marked%20as%20non%2Dresumable%20if%20the%20connection%20is%20not%20closed%20down%20cleanly%2C%20e.g.%20if%20a%20fatal%20error%20occurs%20on%20the%20connection%20or%20SSL_shutdown(3)%20is%20not%20called%20prior%20to%20SSL_free(3).

For completeness, there were two different bits of spec text that led to this OpenSSL behavior, and I only cited one of them. The bits I cited were about unclean shutdown and removed very early. Fatal error lasted all the way to TLS 1.2:

>    Error handling in the TLS Handshake protocol is very simple.  When an
>   error is detected, the detecting party sends a message to the other
>   party.  Upon transmission or receipt of a fatal alert message, both
>   parties immediately close the connection.  Servers and clients MUST
>   forget any session-identifiers, keys, and secrets associated with a
>   failed connection.  Thus, any connection terminated with a fatal
>   alert MUST NOT be resumed.

https://www.rfc-editor.org/rfc/rfc5246.html#section-7.2.2

It wasn't removed until TLS 1.3, in https://github.com/tlswg/tls13-spec/commit/dc668dc1745c458dc510f85084f08458202ddfb7, for the same reasons. In practice, you cannot actually do this consistently, for the reasons I listed above.

Comment (gpshead) [reply Modules/_ssl.c]: Thanks!

> the problem with that commit is that you all turned #if ... #else ... #endif into including both branches instead of just one.

Hah, _that_ explains the code smell here. I'd only been glancing at current version of the code in git blame view to try and understand what this was even for without following the individual commits (git blame doing its job and not exposing history of merely deleted lines)... I agree completely on getting rid of that `_ssl_session_dup` hack.  _(and more, but we need to examine the SOTW and pick an amenable way out of the past API sins in here)_

Comment (davidben) [reply Modules/_ssl.c]: Yeah, definitely makes sense to just start with something minimal (probably this PR) to fix the immediate leak. I think probably the thing to do is:

1. This PR
2. Remove `_ssl_session_dup` and confirm tests indeed now fail in OpenSSL 3.x. My expectation is that they'll still fail, but I've no idea if there were other changes that further complicated this mess.
3. No-op the goofy OpenSSL behavior. Probably something like `SSL_set_quiet_shutdown(ssl); SSL_shutdown(ssl); SSL_free(ssl);` would work? Confirm the tests now pass again.

Comment (pitrou) [reply Modules/_ssl.c]: Calling `SSL_set_shutdown` in the destructor sounds fine to me. Is there a reason to believe that the `SSL_set_quiet_shutdown(ssl); SSL_shutdown(ssl);` dance would be better?

Oh and thanks for the detailed explanation @davidben .

File: Modules/_ssl.c
Comment (pitrou) [inline Modules/_ssl.c]: Hmm, here we _are_ doing the right thing. Should this comment be reworded?

**Comment by pitrou** [comment]
Amusingly, removing the hack doesn't break the tests. Perhaps it was really unnecessary all along?

**Comment by vstinner** [comment]
2 buildbots failed:

> buildbot/ARM64 macOS PR
> buildbot/x86-64 macOS PR

But there are unrelated failure: test_readline_history_file() of test_pyrepl.

**Comment by vstinner** [comment]
I tested the updated PR (only using `SSL_get1_session()`, don't call `_ssl_session_dup()` anymore):

1. It doesn't leak memory anymore: the fix works as expected.
2. It looks **way** faster!

**Comment by pitrou** [comment]
Unsurprisingly, the original `_ssl_session_dup` hack was committed unreviewed:
https://bugs.python.org/issue19500#msg274893

**Comment by davidben** [comment]
> Amusingly, removing the hack doesn't break the tests. Perhaps it was really unnecessary all along?

No, this PR doesn't remove the hack. It's also in `PySSL_set_session`. Removing it from `PySSL_get_session` but keeping it in `PySSL_set_session` is incoherent. Unless there have been some changes since my original analysis, there is definitely a mismatch between what CPython is doing and what OpenSSL expects callers to do.

If anything, the hack _probably_ should have only been in `PySSL_get_session` and not `PySSL_set_session`, so the opposite of what this PR is doing. The effect of the hack is to make a copy of the session that's disconnected from any live connection, because the connection when it dies will invalidate the session[*].

If your goal is to do that disconnect, you should copy the session on get, not set. If you do it on set, you still have a bug if `conn1` is destroyed in between `session = conn1.session` and `conn2.session = session`. Indeed, that's the more likely scenario in real world applications anyway. I suspect CPython's tests just don't cover this.

[*] Again, the only reason it does that is a CPython bug relative to OpenSSL's semantics. OpenSSL's semantics are also a bit goofy and IMO should be reconsidered. But if we take both those problems as read for now.

**Comment by pitrou** [comment]
> No, this PR doesn't remove the hack. It's also in `PySSL_set_session`. Removing it from `PySSL_get_session` but keeping it in `PySSL_set_session` is incoherent.

Oops, thanks for noticing.

> If your goal is to do that disconnect, you should copy the session on get, not set.

_Or_ we just call `SSL_set_shutdown` before `SSL_free`? Would that be enough? It would be nice if we could remove the entire "session dup" hack.

**Comment by davidben** [comment]
> Calling SSL_set_shutdown in the destructor sounds fine to me. Is there a reason to believe that the SSL_set_quiet_shutdown(ssl); SSL_shutdown(ssl); dance would be better?

Either's fine. SSL_set_shutdown is kinda weird in that it scribbles over OpenSSL's internal book-keeping. If you call it wrong, you can even cause the library to get somewhere inconsistent. (I know this because I put guardrails in BoringSSL to stop that from happening, when I couldn't remove the API. 😄)

But if you call it to set bits, rather than clear bits, yeah it's fine. Whichever.

> Or we just call SSL_set_shutdown before SSL_free? Would that be enough? It would be nice if we could remove the entire "session dup" hack.

Yes, that was what I suggested elsewhere. That comment was in the context of how you might do the hack *if you wanted to preserve the hack*. Which you might want as a minimal memory leak fix first. Or if you just want to undo this mess all at once, that's cool too.

**Comment by pitrou** [comment]
Ok, I've tried to removed the session duplication entirely and it works fine locally. I've pushed the change and we can wait for CI.

**Comment by pitrou** [comment]
> 1. It doesn't leak memory anymore: the fix works as expected.
> 2. It looks **way** faster!

Here as well. The session duplication hack was serializing the session object to ASN1, which is apparently quite expensive (looking at the `i2d_SSL_SESSION` source code, there's a *lot* of stuff being serialized).

**Review by vstinner — APPROVED** [review]
LGTM.

I'm not an OpenSSL expert but I trust other developers who were involved in this large change :-)

**Review by gpshead — APPROVED** [review]
I think we're still in "2. Disagree with OpenSSL's policy and effectively turn it off." state with this PR as is due to `SSL_set_shutdown` being used to _mark_ it as "shutdown" so that the OpenSSL implementation does cache it?  There's a nice comment at least. :)

My main goal was get rid of the obvious memory leak. I left the set_session code and obviously gross slow serialize/deserialized stuff in place last night because simply removing that without figuring out the shutdown bits did lead to test failures.  Thanks for picking that up and doing the rest!

**Comment by gpshead** [comment]
Lets wait until @sethmlarson is back online (currently traveling) and can look at this before merging.

It isn't going to make 3.13.0 as it clearly is not a blocker (given 3.10 and 3.11 have always had this flaw and it was only reported this year).  I gather session caching for clients is not super common.

File: Modules/_ssl.c
Comment (davidben) [inline Modules/_ssl.c]: This probably should be either:
```
SSL_set_shutdown(self->ssl, SSL_SENT_SHUTDOWN | SSL_RECEIVED_SHUTDOWN);
```
or:
```
SSL_set_shutdown(self->ssl, SSL_SENT_SHUTDOWN | SSL_get_shutdown(self->ssl));
```

That way if you have already received `close_notify`, you don't clear the bit. I mean, you're immediately destroying the `SSL` anyway, but convincing the library it hasn't received something it has actually received is a little odd. :)

Comment (davidben) [reply Modules/_ssl.c]: (Also because I had put an `assert()` into BoringSSL that callers don't clear bits, because that's weird. 😉 )

**Review by sethmlarson — APPROVED** [review]
Thanks @davidben for all the explanation, this change LGTM.

**Review by SamyDgz — APPROVED** [review]

**Comment by jeffvanvoorst** [comment]
Thanks everyone.  I appreciate the cpython community support.

**Pull request merged**
